### PR TITLE
Add accessible description of current Navigation block state

### LINKED ISSUE #47020: Nav block lacks context about menu currently being edited within the block
As reported in https://github.com/WordPress/gutenberg/issues/46939#issuecomment-1374652632 when entering the Nav block there is no context as to the menu were are editing. It should explain whether it's auto menu or one of the menus we've created. Also perhaps it should announce whether the menu is "saved" or in "draft"?

An accessible `Name` should be provided along the lines of  `Navigation Menu: %%NAV MENU NAME%%` (also consider when the menu is an "auto" menu). 

Related https://github.com/WordPress/gutenberg/issues/47019

### PR DESCRIPTION
**Pull request by getdave**
## What?

Adds accessible description of current state of the Navigation block.

Closes https://github.com/WordPress/gutenberg/issues/47020

## Why?

Current accessible feedback is that the current state of the block is not easily perceivable.

## How?

Adds a description indicating which Navigation Menu is currently being used.

### Saved Menu state

- New Post
- Add Nav block
- Select Navigation block 
- Use `Accessibility` panel in devtools to check description reads "Navigation Menu: %%MENU_NAME%%".

### Unsaved Block State (no menu)

- New Post
- Switch to code view and add the following code to create a Nav block with unsaved inner blocks
```
<!-- wp:navigation -->
<!-- wp:navigation-link {"className":" menu-item menu-item-type-post_type menu-item-object-page","description":"","id":"2","kind":"post-type","label":"Sample Page","opensInNewTab":false,"rel":null,"title":"","type":"page","url":"http://localhost:8888/sample-page/"} /-->
<!-- /wp:navigation -->
```
- Select Navigation block 
- Use `Accessibility` panel in devtools to check description reads "Unsaved Navigation Menu".

### CODE REVIEW

**Review by alexstine — APPROVED** [review]
Nice and simple. Thanks! 👍

**Comment by getdave** [comment]
> Nice and simple. Thanks! 👍

Thanks Alex. Now we have the approval on the general approach from the a11y front let's refine the technical implementation and get ready for merge.

**Comment by scruffian** [comment]
I pushed a commit to this to make AccessibleDescription a wrapping component. The advantage of this approach is that it makes the implementation simpler - the consumer only has to add the component - there's no need to add the additional props to connect them together. If this isn't a good idea feel free to revert my commit.

File: packages/block-library/src/navigation/edit/accessible-description.js [hunk 1]
```diff
@@ -0,0 +1,10 @@
+export default function AccessibleDescription( { id, content } ) {
+	return (
+		<div
+			id={ id }
+			className="wp-block-navigation__description screen-reader-text"
+		>
+			{ content }
```
Comment (getdave) [inline packages/block-library/src/navigation/edit/accessible-description.js]: We should just use `children` as that's built into React.

Then you can just nested the compoentns.

File: packages/block-library/src/navigation/edit/accessible-description.js [hunk 2]
```diff
@@ -0,0 +1,10 @@
+export default function AccessibleDescription( { id, children } ) {
+	return (
+		<div
+			id={ id }
+			className="wp-block-navigation__description screen-reader-text"
```
Comment (getdave) [inline packages/block-library/src/navigation/edit/accessible-description.js]: I don't know if using `screen-reader-text` is legit here. Is that better than `VisuallyHidden`?

Comment (getdave) [reply packages/block-library/src/navigation/edit/accessible-description.js]: I switched to using `VisuallyHidden`.

**Comment by getdave** [comment]
> I pushed a commit to this to make AccessibleDescription a wrapping component.

Can't quite recall how it was before but thanks as what we have now seems fine. I moved it to utilise `React.children` rather than a custom prop for the content and it seems to compose nicely.

What do you think about merging this one?

File: packages/block-library/src/navigation/edit/accessible-description.js (hunk 2 above)
Comment (alexstine) [reply packages/block-library/src/navigation/edit/accessible-description.js]: It's the same thing but `VisuallyHidden` is our React component for it.

**Comment by alexstine** [comment]
Did another accessibility check. Still looking good. 👍

File: packages/block-library/src/navigation/edit/accessible-description.js (hunk 2 above)
Comment (getdave) [reply packages/block-library/src/navigation/edit/accessible-description.js]: As I thought. Let's utilise the standard component then 👍

**Comment by getdave** [comment]
@scruffian This one is ready "as is". I know I overwrote your changes but shall we just merge it and then you can follow up with your additional changes? I'm relatively relaxed either way.

File: packages/block-library/src/navigation/edit/accessible-description.js [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+	isPlaceholder,
+	children,
+} ) {
+	if ( ! children ) {
+		return null;
+	}
+
+	if ( isPlaceholder ) {
+		return children;
+	}
```
Comment (getdave) [inline packages/block-library/src/navigation/edit/accessible-description.js]: I don't like that this component has to know about the concept of a placeholder.

File: packages/block-library/src/navigation/edit/index.js [hunk 4]
```diff
@@ -665,46 +667,59 @@ function Navigation( {
...
-					<UnsavedInnerBlocks
-						createNavigationMenu={ createNavigationMenu }
-						blocks={ uncontrolledInnerBlocks }
-						hasSelection={ isSelected || isInnerBlockSelected }
+			<AccessibleDescription
+				id={ accessibleDescriptionId }
+				content={ __( 'Unsaved Navigation Menu.' ) }
+				isPlaceholder={ isPlaceholder }
+			>
+				<TagName { ...blockProps }>
```
Comment (getdave) [inline packages/block-library/src/navigation/edit/index.js]: IMHO this change makes the code more complex and harder to reason about.

Questions I would ask as a developer seeing this for the first time:

- why is this `AccessibleDescription` component wrapping the main tag?
- how does this work (I now need to dive into the code).
- why does the `AccessibleDescription` need to know about `Placeholder`?
- what is `accessibleDescriptionId` and why isn't it just part of the `AccessibleDescription`?

**Review by getdave — COMMENTED** [review]
I think the last commit makes the code more complex and harder to reason about.

Whilst there is more verbosity in the original approach it is simpler and clearer and I think that outweighs any benefits derived from the approach in this latest commit.

My opinion is that we revert to the previous state, merge and _then_ follow up _if_ we feel this new approach still has merit.

**Comment by scruffian** [comment]
Thanks for the review Dave. I have removed my commit, so I think this is good to go!

**Review by scruffian — APPROVED** [review]

**Comment by getdave** [comment]
> Thanks for the review Dave. I have removed my commit, so I think this is good to go!

Thank you for being open to feedback. I appreciate it 🙇

**Pull request merged**